spark-bigquery-pushdown/spark-3.1-bigquery-pushdown_2.12/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark31BigQueryStrategy.scala

```diff
@@ -1,11 +1,31 @@
 package com.google.cloud.spark.bigquery.pushdowns
-import com.google.cloud.bigquery.connector.common.BigQueryPushdownUnsupportedException
+import com.google.cloud.spark.bigquery.SupportsQueryPushdown
+import com.google.cloud.spark.bigquery.pushdowns.SparkBigQueryPushdownUtil.getTableNameFromOptions
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanRelation
+import scala.collection.JavaConverters._
 
 class Spark31BigQueryStrategy(expressionConverter: SparkExpressionConverter, expressionFactory: SparkExpressionFactory, sparkPlanFactory: SparkPlanFactory)
   extends BigQueryStrategy(expressionConverter, expressionFactory, sparkPlanFactory) {
   override def generateQueryFromPlanForDataSourceV2(plan: LogicalPlan): Option[BigQuerySQLQuery] = {
-    throw new BigQueryPushdownUnsupportedException("Query pushdown for Spark 3.1 DataSourceV2 connector has not been implemented yet")
+    // DataSourceV2ScanRelation is the relation that is used in the Spark 3.1 DatasourceV2 connector
+    plan match {
+      case scanRelation@DataSourceV2ScanRelation(_, _, _) =>
+        // Get the scan and cast it to SupportsQueryPushdown to get the BigQueryRDDFactory
+        val scan = scanRelation.scan.asInstanceOf[SupportsQueryPushdown]
```

Comment:
What happens if a scan does not implement SupportsQueryPushdown ? A FileScan for example?

Reply:
Good question. I updated the code to pattern match on `SupportsQueryPushdown`. 

If a scan does not implement SupportsQueryPushdown, then `None` will be returned from this method and we will resort to our regular non query pushdown code path